Create rule S7128: "strlen" should not be called on incremented pointers (CPP-5660) (#4420)
SonarSource/rspec: rules/S7128/cfamily/rule.adoc

> This rule raises the issue of calls `strlen(ptr + 1)`, for which the intent is unclear.

Reviewer: ```suggestion
Calling `strlen(ptr + 1)` looks very similar to the more common pattern `strlen(ptr) + 1` and has unclear intent.
```

== Why is this an issue?

The call `strlen(ptr) + 1` is commonly used to compute the memory required to copy a string:
the one is needed to allocate space for a null-terminator.

The very similar call `strlen(ptr + 1)` that differs by placement of the right parenthesis,
produces vastly different results:

 * non-empty string computes the length of string without the first character (reduced by one)

>  * if the string is empty, traverse memory after it until the null terminator (`'\0'`) is found.

Reviewer: ```suggestion
 * if the string is empty, it traverses neighboring memory until a null terminator (`'\0'`) is found.
```

The latter may lead to undefined behavior.

> Due above, the `strlen(ptr + 1)` calls are likely to be unintended
> and result from the type. This rule raises issues on such calls.

Reviewer: ```suggestion
In consequence, calls to `strlen(ptr + 1)` are likely to be unintended
and the result of a typo. This rule raises issues on such calls.
```

=== Code examples

==== Noncompliant code example

> [source,c]

Reviewer: ```suggestion
[source,c,diff-id=1,diff-type=noncompliant]
```

----
size_t size = strlen(ptr + 1); // Noncompliant
----

==== Compliant solution

If the use of `strlen(ptr + 1)` was an typo:

> [source,c]

Reviewer: ```suggestion
[source,c,diff-id=1,diff-type=compliant]
```

----
size_t size = strlen(ptr) + 1; // Compliant
----

> In case when the `strlen(ptr + 1)` was intentional, to preserve the same behavior, you may increment the pointer before the call.

Reviewer: Or we could propose to use `strlen(ptr) - 1`, which can be protected with
```cpp
std::size fullSize = strlen(ptr);
std::size size = 0;
if (fullSize > 0) {
  size = fullSize - 1;
}
```

Reviewer 2: This is not the same behavior, if you have a memory filled with null-termined strings, and you are rebuilding it. In that case `ptr` may be pointing out to `'\0'` and `next` is next string.

I do not want to go into that level of detail here, so decide to present option that for sure does not change meaning

[source,c]
----
char const* next = ptr + 1;
size_t size = strlen(next); // Compliant
----

This will convey the intent more clearly and avoid any uncertainty regarding the intent.

=== What is the potential impact?

If `strlen(ptr + 1)` is accidentally used instead of `strlen(ptr) + 1`, it may have a wide range of effects.
For example, it can lead to buffer overflow if the operation was used to compute the required memory:

[source,c]
----
char* duplicate(char const* source) {
  char result = (char*)malloc(strlen(source + 1)); // Should be malloc(strlen(source) + 1))
  strcpy(result, source);  // Writes two characters outside of "result" buffer
  return result;
}
----

> Buffer overflow, as other undefined behavior, has a wide range of effects.

Reviewer: ```suggestion
Buffer overflows, as other undefined behavior, have a wide range of effects.
```
